### PR TITLE
perf + tests: improve efficiency for tsc + add test for workflow_1 (ic_usd, tsc_usd)

### PR DESCRIPTION
**Pull request by anujsinha3**
#51  #62  #63 #64 

test: add test for workflow1 for ic+usd
test: add test for workflow2 for tsc+usd
test: update constants for test assertion
perf: use dict to store distance values to avoid re-computation for duplicate points
fix: update stay duration grouping logic
fix: trace segmentation stay_lat, stay_lon for single group.

On M1 Pro 16GB 6+2 Cores for 967794 trace rows for 5994 unique users
Before the changes: tc+usd took approx ~30 secs, post change it takes around ~28 secs.

### CODE REVIEW

File: src/mawpy/steps/trace_segmentation_clustering.py [hunk 1]
```diff
@@ -21,9 +21,15 @@ def _get_diameter_constraint_exceed_index(starting_index: int, point_to_check: i
...
+    # Map to store distance values to avoid re-computation for duplicate points
+    distance_map = {}
     for i in range(starting_index, point_to_check):
-        if spatial_constraint < distance(latitudes_list[i], longitudes_list[i],
-                                         latitudes_list[point_to_check], longitudes_list[point_to_check]):
+        point_key = (latitudes_list[i], longitudes_list[i])
+        if point_key not in distance_map:
+            distance_map[point_key] = distance(latitudes_list[i], longitudes_list[i],
+                                               latitudes_list[point_to_check], longitudes_list[point_to_check])
+        if spatial_constraint < distance_map[point_key]:
```
Comment (carlosgjs) [inline src/mawpy/steps/trace_segmentation_clustering.py:32]: This is likely very minor, but you can avoid doing the map lookup twice
```suggestion
        d = distance_map.get(point_key, -1)
        if d == -1:
            d = distance(latitudes_list[i], longitudes_list[i],
                                               latitudes_list[point_to_check], longitudes_list[point_to_check])
            distance_map[point_key] = d
        if spatial_constraint < d:
```

File: tests/workflows/test_workflow1.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+from mawpy.constants import (USER_ID, UNIX_START_T, UNIX_START_DATE, ORIG_LAT, ORIG_LONG, ORIG_UNC, STAY_LAT, STAY_LONG,
+                             STAY_UNC, STAY, STAY_DUR)
+from mawpy.steps.incremental_clustering import incremental_clustering
+from mawpy.steps.update_stay_duration import update_stay_duration
+from mawpy.workflows.workflow1 import workflow1
+
+
+def test_workflow1_ic_usd():
+    print(os.getcwd())
```
Comment (carlosgjs) [inline tests/workflows/test_workflow1.py:15]: logger? or remove?

File: tests/workflows/test_workflow1.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    with TemporaryDirectory() as tempdir:
+        input_file = os.path.dirname(__file__) + '/../resources/test_input.csv'
+        output_file = os.path.join(tempdir, 'test_output_workflow1.csv')
+
+        # Prepare arguments similar to command-line arguments
```
Comment (carlosgjs) [inline tests/workflows/test_workflow1.py:20]: why is this needed? you could just pass the variables

File: tests/workflows/test_workflow1.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        pd.testing.assert_frame_equal(actual_output_df, expected_output_df)
+
+        # List of columns expected in the actual_output_df
+        expected_output_columns = [USER_ID, UNIX_START_T, UNIX_START_DATE,
+                                   ORIG_LAT, ORIG_LONG, ORIG_UNC,
+                                   STAY_LAT, STAY_LONG, STAY_UNC,
+                                   STAY, STAY_DUR]
+
+        # Check if all the expected columns are present in the workflow output columns
+        assert len(list(set(expected_output_columns) & set(actual_output_df.columns))) == len(expected_output_columns)
```
Comment (carlosgjs) [inline tests/workflows/test_workflow1.py:54]: doesn't the `assert_frame_equal` cover this?

File: tests/workflows/test_workflow1.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+from mawpy.constants import (USER_ID, UNIX_START_T, UNIX_START_DATE, ORIG_LAT, ORIG_LONG, ORIG_UNC, STAY_LAT, STAY_LONG,
+                             STAY_UNC, STAY, STAY_DUR)
+from mawpy.steps.incremental_clustering import incremental_clustering
+from mawpy.steps.update_stay_duration import update_stay_duration
+from mawpy.workflows.workflow1 import workflow1
+
+
+def test_workflow1_ic_usd():
+    print(os.getcwd())
+    with TemporaryDirectory() as tempdir:
```
Comment (carlosgjs) [inline tests/workflows/test_workflow1.py:16]: you could use the tempdir fixture: https://docs.pytest.org/en/6.2.x/tmpdir.html

File: tests/workflows/test_workflow1.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                                   STAY, STAY_DUR]
+
+        # Check if all the expected columns are present in the workflow output columns
+        assert len(list(set(expected_output_columns) & set(actual_output_df.columns))) == len(expected_output_columns)
+
+        # Check if output file was written successfully
+        assert os.path.exists(args.output_file)
+
+        # Destroy test output file after test completed
+        os.remove(args.output_file)
```
Comment (carlosgjs) [inline tests/workflows/test_workflow1.py:60]: Should not be needed since you are using a temp dir

**Review by carlosgjs — APPROVED** [review]
See inline suggestions

File: tests/workflows/test_workflow1.py (hunk 4 above)
Comment (anujsinha3) [reply tests/workflows/test_workflow1.py:54]: assert frame equal checks equality of dataframes returned by workflow1 vs the df output of combination of IC_USD steps. It may happen that those are same but don't have all the columns listed above. So this check becomes helpful in that case.

File: tests/workflows/test_workflow1.py (hunk 3 above)
Comment (anujsinha3) [reply tests/workflows/test_workflow1.py]: Can you please elaborate a bit more?

do you mean directly passing
os.path.dirname(__file__) + '/../resources/test_input.csv'
os.path.join(tempdir, 'test_output_workflow1.csv')
